src/backend/joanie/core/models/products.py

```diff
@@ -261,16 +262,23 @@ def validate(self):
         """
 
         # Enroll user to course run that are the only one course run of the course
-        courses = self.target_courses.annotate(
+        courses_with_one_course_run = self.target_courses.annotate(
             course_runs_count=models.Count("course_runs")
         ).filter(course_runs_count=1)
-        course_runs = courses_models.CourseRun.objects.filter(course__in=courses)
+
+        opened_courses = [
+            course
+            for course in courses_with_one_course_run
+            if course.state["priority"] <= CourseState.ARCHIVED_OPEN
```

Comment:
did you count the number of queries? Maybe prefetch related would help?

Reply:
Currently, there are 19 db queries executed. Use `prefetch_related` does not reduce the number of queries in this case.

On the other hand, I tried to retrieve course runs in only one queryset. I succeeded with the following one :
```py
courses_with_one_course_run = self.target_courses.annotate(
                course_runs_count=models.Count("course_runs")
            ).filter(
                Q(course_runs__enrollment_end__gt=timezone.now())
                | Q(course_runs__enrollment_end__isnull=True),
                course_runs__enrollment_start__lte=timezone.now(),
                course_runs_count=1,
            )
```

In this case, `order.validate()` executes 15 db queries. WDYT?

Reply:
After running some tests I agree that we can leave it as it is. 
In fact 11 of the 4 queries are coming from the enrollment's save method calling `full_clean` and testing a lot of things. So this is where some optimization could be done but not in the validate :man_shrugging: 
Maybe some of the tests in full_clean could be translated to db constraints.
This being said, we don't really need to optimize this code as it is only called when we get an order... so sadly not so many times per minute...

Reply:
I created issue https://github.com/openfun/joanie/issues/74#issue-1101520420 to remind to optimize db queries later.